README.md
# Gatsby Plugin Azure Search

Gatsby plugin to ingest data into Microsoft Azure Cognitive Search.

## Overview

There are 2 parts of implementing a search feature for your site. The indexing (writing) part and the querying (reading) part. The indexing part push data into some search back end, the querying part reads from the search back end via its APIs then presents the result on the front end.

This plugin handles only the indexing part, using [Azure Cognitive Search](https://azure.microsoft.com/en-us/services/search/) as the back end service.

## How to use

Install the plugin.

```bash
npm install --save gatsby-plugin-azure-search dotenv
```

Create an `.env` file in the root directory of gatsby, same level as `gatsby-config.js`. Do not commit this file.

```
AZURE_SEARCH_SERVICE_NAME=X
AZURE_SEARCH_ADMIN_KEY=X // Note: we need "admin key" instead of "query key"
```

It is recommended to use dotenv for security reasons. If your Gatsby repository is always private, you can hard-code the credentials in `gatsby-config.js` at your own risk.

Add configuration in gatsby-config.js:

```javascript
require('dotenv').config({
  path: `.env`,
});

module.exports = {
  plugins: [
    {
      resolve: `gatsby-plugin-azure-search`,
      options: {
        verbose: false, // default: false
        serviceName: process.env.AZURE_SEARCH_SERVICE_NAME, // required
        apiKey: process.env.AZURE_SEARCH_ADMIN_KEY, // required
        indexConfig: { // required. refer to azure documentation
          name: ``, // required. the plugin upserts the index, no need to create it in advance
          fields: [], // required
          suggesters: [], // optional
          scoringProfiles: [], // optional
          analyzers: [], // optional
          charFilters: [], // optional
          tokenizers: [], // optional
          tokenFilters: [], // optional
          defaultScoringProfile: '', // optional
          corsOptions: { // optional
            allowedOrigins: [], 
            maxAgeInSeconds: 300, 
          },
          encryptionKey: {}, // optional
        },
        queries: [], // required. details in next section.
      },
    }
  ]
}
```

For `indexConfig`, refer to the official Azure [documentation](https://docs.microsoft.com/en-us/azure/search/search-what-is-an-index#index-attributes).

You may also refer to `example` folder for how I configured the plugin in my blog.

Assume the configuration is correct, your gatsby site should be indexed to Azure search every time it is built. You can verify the generated search index in Azure console.

### Query Configuration

You can provide multiple graphql queries for ingestion, but for now all generated documents will be ingested into the same index.

Each query object specifies a required graphql `query` and an optional `transformer` function.

The transformer function operates on the raw graphql query output as an array, and returns the transformed indexable documents as an array.

The transformed document must match exactly as defined in the index's `fields` configuration. Including extra keys will result in bad request errors from Azure.

Sample query:

```javascript
const sampleQuery = {
  query: `{
    allWordpressPost(
      filter: {
        status: { eq: "publish" }
      }
    ) {
      edges {
        node {
          slug
          date
          title
          content
          excerpt
          categories {
            name
          }
          tags {
            name
          }
        }
      }
    }
  }`,
  transformer: ({ data }) => {
    return data.allWordpressPost.edges.map(edge => {
      return {
        ...edge.node,
        permalink: `https://artifact.me/${edge.node.slug}`,
        content: (edge.node.content || '').replace(/(<([^>]+)>)/ig,""),
        excerpt: (edge.node.excerpt || '').replace(/(<([^>]+)>)/ig,"").substring(0, 200) + '...',
        categories: (edge.node.categories || []).map(c => c.name),
        tags: (edge.node.tags || []).map(t => t.name),
      };
    });
  },
};
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.13](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.12...v0.0.13) (2023-01-30)

- Update dependencies

### [0.0.12](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.11...v0.0.12) (2022-01-22)

### [0.0.11](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.9...v0.0.11) (2021-10-04)

### [0.0.10](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.9...v0.0.10) (2021-10-04)

### [0.0.9](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.8...v0.0.9) (2021-03-07)

### [0.0.8](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.7...v0.0.8) (2021-01-17)

### [0.0.7](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.6...v0.0.7) (2020-10-05)

- Add an example `gatsby-config.js`.
- Update dependencies.

### [0.0.6](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.5...v0.0.6) (2020-06-07)

### [0.0.5](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.4...v0.0.5) (2020-05-16)

### [0.0.4](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.3...v0.0.4) (2020-05-13)

### [0.0.3](https://github.com/artchen/gatsby-plugin-azure-search/compare/v0.0.2...v0.0.3) (2020-05-12)

### 0.0.2 (2020-05-12)


gatsby-node.js
const axios = require('axios');

const API_VERSION = '2019-05-06';

/**
 * Get common header to call Azure search REST APIs.
 * @param apiKey azure search admin key
 * @returns {{'api-key': *, 'Content-Type': string}}
 */
const getHeaders = ({ apiKey }) => {
  return {
    'api-key': apiKey,
    'Content-Type': 'application/json',
  };
};

/**
 * Delete a search index.
 * @param serviceName azure search service name
 * @param apiKey azure search admin key
 * @param indexConfig specified in gatsby-config.js
 * @param verbose
 * @param reporter
 * @returns {Promise<*>}
 */
const deleteIndex = async ({
  serviceName,
  apiKey,
  indexConfig,
  verbose = false,
  reporter,
}) => {
  const url = `https://${serviceName}.search.windows.net/indexes/${indexConfig.name}?api-version=${API_VERSION}`;
  const headers = getHeaders({ apiKey });
  if (verbose) {
    reporter.info(`Input to delete index:`);
    console.log(`url:`, url);
    console.log(`indexName:`, indexConfig.name);
    console.log(`headers:`, headers);
  }
  try {
    const res = await axios({
      method: 'delete',
      url,
      headers,
    });
    reporter.info(`Deleted index: ${indexConfig.name}`);
    return Promise.resolve(res);
  } catch (e) {
    reporter.warn(`Faield to delete index: ${e.message}. But it might be OK.`);
    if (verbose) {
      console.log(`Error response:`, e.response.data);
    }
    return Promise.resolve();
  }
};

/**
 * Create a search index.
 * @param serviceName azure search service name
 * @param apiKey azure search admin key
 * @param indexConfig specified in gatsby-config.js
 * @param verbose
 * @param reporter
 * @returns {Promise<*>}
 */
const createIndex = async ({
  serviceName,
  apiKey,
  indexConfig,
  verbose = false,
  reporter,
}) => {
  const url = `https://${serviceName}.search.windows.net/indexes/${indexConfig.name}?api-version=${API_VERSION}`;
  const headers = getHeaders({ apiKey });
  if (verbose) {
    reporter.info(`Input to create index:`);
    console.log(`url:`, url);
    console.log(`indexConfig:`, indexConfig);
    console.log(`headers:`, headers);
  }
  try {
    const res = await axios({
      method: 'put',
      url,
      headers,
      data: indexConfig,
    });
    reporter.info(`Created index: ${indexConfig.name}`);
    return Promise.resolve(res);
  } catch (e) {
    reporter.error(`Faield to create index: ${e.message}`);
    if (verbose) {
      console.log(`Error response:`, e.response.data);
    }
    return Promise.reject(e);
  }
};

/**
 * Index an array of documents to Azure.
 * @param serviceName azure search service name
 * @param apiKey azure search admin key
 * @param indexName indexConfig.name
 * @param docs documents to index
 * @param verbose
 * @param reporter
 * @returns {*}
 */
const indexDocuments = async ({
  serviceName,
  apiKey,
  indexName,
  docs,
  verbose,
  reporter,
}) => {
  if(!docs || !docs.length) {
    reporter.warn(`Skipped: No docs to reindex`)
    return Promise.resolve({})
  }
  const url = `https://${serviceName}.search.windows.net/indexes/${indexName}/docs/index?api-version=${API_VERSION}`;
  const body = {
    value: docs,
  };
  const headers = getHeaders({ apiKey });
  if (verbose) {
    reporter.info(`Input to index document:`);
    console.log(`url:`, url);
    reporter.log(`headers:`, headers);
  }
  try {
    const res = await axios({
      method: 'post',
      url,
      headers,
      data: body,
    });
    reporter.info(`Indexed documents to ${indexName}`);
    if (verbose) {
      reporter.log(`Response data: `, JSON.stringify(res.data));
    }
    return Promise.resolve(res);
  } catch (e) {
    reporter.error(`Faield to index documents: ${e.message}`);
    if (verbose) {
      console.log(`Error response:`, e.response.data);
    }
    return Promise.reject(e);
  }
};

/**
 * Default no-op transformer function
 */
const identity = (docs) => docs;

/**
 * Run graphql query then index documents to Azure.
 * @param graphql
 * @param queryIndex
 * @param query the query config specified in gatsby-config.js
 * @param transformer the transformer function specified in gatsby-config.js
 * @param serviceName azure search service name
 * @param apiKey azure search admin key
 * @param indexName indexConfig.name
 * @param verbose
 * @param reporter
 * @returns {Promise<*>}
 */
const doQuery = async ({
  graphql,
  queryIndex,
  query,
  transformer = identity,
  serviceName,
  apiKey,
  indexName,
  verbose,
  reporter,
}) => {
  if (!query) {
    reporter.panic(`Please specify "query"`);
  }

  reporter.info(`Query ${queryIndex}: running graphql query`);
  const result = await graphql(query);

  if (result.errors) {
    const errMsg = `Failed to run graphql query`;
    reporter.panic(errMsg, result.errors);
    return Promise.reject(new Error(errMsg));
  }

  reporter.info(`Query ${queryIndex}: running transformer`);
  const docs = await transformer(result);

  reporter.info(`Query ${queryIndex}: generated ${docs.length} documents`);
  reporter.info(`Query ${queryIndex}: indexing documents`);
  return indexDocuments({ serviceName, apiKey, indexName, docs, verbose, reporter });
};

// Gatsby API
exports.onPostBuild = async function (
  { graphql, reporter },
  { serviceName, apiKey, indexConfig, queries, verbose = false }
) {
  const activity = reporter.activityTimer('Index to Azure Search');
  activity.start();

  try {
    reporter.info(`Rebuild index ${indexConfig.name}`);
    await deleteIndex({ serviceName, apiKey, indexConfig, verbose, reporter });
    await createIndex({ serviceName, apiKey, indexConfig, verbose, reporter });

    reporter.info(`${queries.length} queries to index`);
    const jobs = queries.map((query, queryIndex) =>
      doQuery({
        ...query,
        queryIndex,
        graphql,
        activity,
        serviceName,
        apiKey,
        indexName: indexConfig.name,
        verbose,
        reporter,
      })
    );

    await Promise.all(jobs);
  } catch (e) {
    reporter.panic(`Failed to index to Azure Search.`);
  } finally {
    activity.end();
  }
};


example/gatsby-config.example.js
require('dotenv').config({
  path: `.env`,
});

const gatsbyPluginAzureSearch = {
  resolve: `gatsby-plugin-azure-search`,
  options: {
    verbose: true,
    serviceName: process.env.AZURE_SEARCH_SERVICE_NAME,
    apiKey: process.env.AZURE_SEARCH_ADMIN_KEY,
    indexConfig: {
      name: `otakism`,
      corsOptions: {
        allowedOrigins: ['*'],
        maxAgeInSeconds: 300,
      },
      fields: [
        {
          name: 'slug',
          type: 'Edm.String',
          searchable: false,
          filterable: false,
          retrievable: true,
          sortable: true,
          facetable: false,
          key: true,
        },
        {
          name: 'title',
          type: 'Edm.String',
          searchable: true,
          filterable: false,
          retrievable: true,
          sortable: true,
          facetable: false,
          key: false,
          analyzer: 'zh-Hans.lucene',
        },
        {
          name: 'date',
          type: 'Edm.DateTimeOffset',
          searchable: false,
          filterable: false,
          retrievable: true,
          sortable: true,
          facetable: false,
          key: false,
        },
        {
          name: 'content',
          type: 'Edm.String',
          searchable: true,
          filterable: false,
          retrievable: true,
          sortable: false,
          facetable: false,
          key: false,
          analyzer: 'zh-Hans.lucene',
        },
        {
          name: 'excerpt',
          type: 'Edm.String',
          searchable: true,
          filterable: false,
          retrievable: true,
          sortable: false,
          facetable: false,
          key: false,
          analyzer: 'zh-Hans.lucene',
        },
        {
          name: 'permalink',
          type: 'Edm.String',
          searchable: false,
          filterable: false,
          retrievable: true,
          sortable: false,
          facetable: false,
          key: false,
        },
        {
          name: 'categories',
          type: 'Collection(Edm.String)',
          searchable: true,
          filterable: true,
          retrievable: true,
          sortable: false,
          facetable: false,
          key: false,
        },
        {
          name: 'tags',
          type: 'Collection(Edm.String)',
          searchable: true,
          filterable: true,
          retrievable: true,
          sortable: false,
          facetable: false,
          key: false,
        },
      ],
    },
    queries: [
      {
        query: `{
            allWordpressPage(
              filter: {
                status: { eq: "publish" }
              }
            ) {
              edges {
                node {
                  slug
                  title
                  date
                  content
                  excerpt
                }
              }
            }
          }`,
        transformer: ({ data }) => {
          return data.allWordpressPage.edges.map((edge, index) => {
            return {
              ...edge.node,
              permalink: `${siteMetadata.siteUrl}/${edge.node.slug}`,
              content: (edge.node.content || '').replace(/(<([^>]+)>)/gi, ''),
              excerpt: (edge.node.excerpt || edge.node.content || '')
                .replace(/(<([^>]+)>)/gi, '')
                .substring(0, 200),
              categories: [],
              tags: [],
            };
          });
        },
      },
      {
        query: `{
            allWordpressPost(
              filter: {
                status: { eq: "publish" }
              }
            ) {
              edges {
                node {
                  slug
                  date
                  title
                  content
                  excerpt
                  categories {
                    name
                  }
                  tags {
                    name
                  }
                }
              }
            }
          }`,
        transformer: ({ data }) => {
          return data.allWordpressPost.edges.map((edge) => {
            return {
              ...edge.node,
              permalink: `${siteMetadata.siteUrl}/${edge.node.slug}`,
              content: (edge.node.content || '').replace(/(<([^>]+)>)/gi, ''),
              excerpt: (edge.node.content || '')
                .replace(/(<([^>]+)>)/gi, '')
                .substring(0, 200),
              categories: (edge.node.categories || []).map((c) => c.name),
              tags: (edge.node.tags || []).map((t) => t.name),
            };
          });
        },
      },
    ],
  },
};

module.exports = {
  plugins: [
    gatsbyPluginAzureSearch,
  ],
};
